### ISSUE TITLE
Deadlock via JavaChangeQueueFiller

### ISSUE THREAD

**Issue by iloveeclipse** [eclipse/xtext#1460]
XText 2.16, deadlock just after IDE startup.

The problem is that `"Worker-0: Build"` holds Xtext lock which and waits for workspace lock, and `"Worker-22: Updating workspace"` thread holds workspace lock and waits for XText lock.

```
"Worker-22: Updating workspace" #81 prio=5 os_prio=0 tid=0x00007ffe5c001000 nid=0x16ff2 waiting for monitor entry [0x00007ffe91794000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at org.eclipse.xtext.builder.builderState.AbstractBuilderState.load(AbstractBuilderState.java:59)
	- waiting to lock <0x000000074af06250> (a org.eclipse.xtext.builder.clustering.ClusteringBuilderState)
	at org.eclipse.xtext.builder.builderState.AbstractBuilderState.ensureLoaded(AbstractBuilderState.java:69)
	at org.eclipse.xtext.builder.builderState.AbstractBuilderState.getExportedObjects(AbstractBuilderState.java:199)
	at org.eclipse.xtext.common.types.ui.notification.DeltaConverter.isDerived(DeltaConverter.java:370)
	at org.eclipse.xtext.common.types.ui.notification.DeltaConverter.isDerived(DeltaConverter.java:362)
	at org.eclipse.xtext.common.types.ui.notification.DeltaConverter.convertChangedCompilationUnit(DeltaConverter.java:191)
	at org.eclipse.xtext.common.types.ui.notification.DeltaConverter.convertCompilationUnit(DeltaConverter.java:167)
	at org.eclipse.xtext.common.types.ui.notification.DeltaConverter.convertCompilationUnits(DeltaConverter.java:467)
	at org.eclipse.xtext.common.types.ui.notification.DeltaConverter.convertCompilationUnits(DeltaConverter.java:471)
	at org.eclipse.xtext.common.types.ui.notification.DeltaConverter.convertCompilationUnits(DeltaConverter.java:471)
	at org.eclipse.xtext.common.types.ui.notification.DeltaConverter.convertCompilationUnits(DeltaConverter.java:471)
	at org.eclipse.xtext.common.types.ui.notification.DeltaConverter.convertCompilationUnits(DeltaConverter.java:471)
	at org.eclipse.xtext.common.types.ui.notification.DeltaConverter.convert(DeltaConverter.java:148)
	at org.eclipse.xtext.builder.impl.javasupport.JavaChangeQueueFiller.elementChanged(JavaChangeQueueFiller.java:38)
	at org.eclipse.jdt.internal.core.DeltaProcessor$3.run(DeltaProcessor.java:1738)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
	at org.eclipse.jdt.internal.core.DeltaProcessor.notifyListeners(DeltaProcessor.java:1726)
	at org.eclipse.jdt.internal.core.DeltaProcessor.firePostChangeDelta(DeltaProcessor.java:1559)
	at org.eclipse.jdt.internal.core.DeltaProcessor.fire(DeltaProcessor.java:1535)
	at org.eclipse.jdt.internal.core.DeltaProcessor.notifyAndFire(DeltaProcessor.java:2256)
	at org.eclipse.jdt.internal.core.DeltaProcessor.resourceChanged(DeltaProcessor.java:2146)
	at org.eclipse.jdt.internal.core.DeltaProcessingState.resourceChanged(DeltaProcessingState.java:477)
	at org.eclipse.core.internal.events.NotificationManager$1.run(NotificationManager.java:300)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
	at org.eclipse.core.internal.events.NotificationManager.notify(NotificationManager.java:290)
	at org.eclipse.core.internal.events.NotificationManager.broadcastChanges(NotificationManager.java:153)
	at org.eclipse.core.internal.resources.Workspace.broadcastPostChange(Workspace.java:379)
	at org.eclipse.core.internal.resources.Workspace.endOperation(Workspace.java:1502)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2308)
	at org.eclipse.core.internal.events.NotificationManager$NotifyJob.run(NotificationManager.java:44)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
```

```
"Worker-0: Build" #31 prio=5 os_prio=0 tid=0x00007ffff0ba4800 nid=0x16fa4 in Object.wait() [0x00007fff40de8000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	at org.eclipse.core.internal.jobs.Semaphore.acquire(Semaphore.java:42)
	- locked <0x0000000716d40370> (a org.eclipse.core.internal.jobs.Semaphore)
	at org.eclipse.core.internal.jobs.OrderedLock.doAcquire(OrderedLock.java:173)
	at org.eclipse.core.internal.jobs.OrderedLock.acquire(OrderedLock.java:109)
	at org.eclipse.core.internal.jobs.OrderedLock.acquire(OrderedLock.java:85)
	at org.eclipse.core.internal.resources.WorkManager.endUnprotected(WorkManager.java:190)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2307)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2322)
	at org.eclipse.xtext.ui.resource.Storage2UriMapperJavaImpl.doInitializeCache(Storage2UriMapperJavaImpl.java:667)
	at org.eclipse.xtext.ui.resource.Storage2UriMapperJavaImpl.initializeCache(Storage2UriMapperJavaImpl.java:575)
	at org.eclipse.xtext.ui.resource.Storage2UriMapperJavaImpl.initializeCache(Storage2UriMapperJavaImpl.java:534)
	at org.eclipse.xtext.ui.resource.Storage2UriMapperImpl.initializeCache(Storage2UriMapperImpl.java:162)
	at org.eclipse.xtext.builder.builderState.AbstractBuilderState.load(AbstractBuilderState.java:62)
	- locked <0x000000074af06250> (a org.eclipse.xtext.builder.clustering.ClusteringBuilderState)
	at org.eclipse.xtext.builder.builderState.AbstractBuilderState.ensureLoaded(AbstractBuilderState.java:69)
	at org.eclipse.xtext.builder.builderState.AbstractBuilderState.isEmpty(AbstractBuilderState.java:217)
	at org.eclipse.xtext.ui.shared.internal.ListenerRegistrar.initialize(ListenerRegistrar.java:44)
	at org.eclipse.xtext.ui.shared.internal.EagerContributionInitializer.initialize(EagerContributionInitializer.java:31)
	at org.eclipse.xtext.ui.shared.internal.Activator.start(Activator.java:214)
	at org.eclipse.osgi.internal.framework.BundleContextImpl$3.run(BundleContextImpl.java:782)
	at org.eclipse.osgi.internal.framework.BundleContextImpl$3.run(BundleContextImpl.java:1)
	at java.security.AccessController.doPrivileged(Native Method)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.startActivator(BundleContextImpl.java:775)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.start(BundleContextImpl.java:732)
	at org.eclipse.osgi.internal.framework.EquinoxBundle.startWorker0(EquinoxBundle.java:1005)
	at org.eclipse.osgi.internal.framework.EquinoxBundle$EquinoxModule.startWorker(EquinoxBundle.java:357)
	at org.eclipse.osgi.container.Module.doStart(Module.java:589)
	at org.eclipse.osgi.container.Module.start(Module.java:457)
	at org.eclipse.osgi.framework.util.SecureAction.start(SecureAction.java:471)
	at org.eclipse.osgi.internal.hooks.EclipseLazyStarter.postFindLocalClass(EclipseLazyStarter.java:117)
	at org.eclipse.osgi.internal.loader.classpath.ClasspathManager.findLocalClass(ClasspathManager.java:557)
	at org.eclipse.osgi.internal.loader.ModuleClassLoader.findLocalClass(ModuleClassLoader.java:331)
	at org.eclipse.osgi.internal.loader.BundleLoader.findLocalClass(BundleLoader.java:395)
	at org.eclipse.osgi.internal.loader.BundleLoader.findClassInternal(BundleLoader.java:473)
	at org.eclipse.osgi.internal.loader.BundleLoader.findClass(BundleLoader.java:422)
	at org.eclipse.osgi.internal.loader.BundleLoader.findClass(BundleLoader.java:414)
	at org.eclipse.osgi.internal.loader.ModuleClassLoader.loadClass(ModuleClassLoader.java:153)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at org.eclipse.osgi.internal.framework.EquinoxBundle.loadClass(EquinoxBundle.java:612)
	at org.eclipse.core.internal.registry.osgi.RegistryStrategyOSGI.createExecutableExtension(RegistryStrategyOSGI.java:196)
	at org.eclipse.core.internal.registry.ExtensionRegistry.createExecutableExtension(ExtensionRegistry.java:934)
	at org.eclipse.core.internal.registry.ConfigurationElement.createExecutableExtension(ConfigurationElement.java:246)
	at org.eclipse.core.internal.registry.ConfigurationElementHandle.createExecutableExtension(ConfigurationElementHandle.java:63)
	at org.eclipse.core.internal.events.BuildManager.instantiateBuilder(BuildManager.java:1012)
	at org.eclipse.core.internal.events.BuildManager.initializeBuilder(BuildManager.java:951)
	at org.eclipse.core.internal.events.BuildManager.getBuilder(BuildManager.java:633)
	at org.eclipse.core.internal.events.BuildManager.getBuilder(BuildManager.java:673)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:261)
	at org.eclipse.core.internal.events.BuildManager$1.run(BuildManager.java:316)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:319)
	at org.eclipse.core.internal.events.BuildManager.basicBuildLoop(BuildManager.java:371)
	at org.eclipse.core.internal.events.BuildManager.build(BuildManager.java:392)
	at org.eclipse.core.internal.resources.Workspace.buildInternal(Workspace.java:504)
	at org.eclipse.core.internal.resources.Workspace.build(Workspace.java:404)
	at org.eclipse.oomph.setup.internal.core.SetupTaskPerformer$5.run(SetupTaskPerformer.java:3379)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
```

**Comment by iloveeclipse** [comment]
@cdietrich FYI

**Comment by cdietrich** [comment]
@iloveeclipse does this happen with 2.17.1 too?
cc @szarnekow

**Comment by cdietrich** [comment]
@iloveeclipse do you see where Storage2UriMapperJavaImpl is waiting for?

**Comment by iloveeclipse** [comment]
We do not have 2.17 in use yet. 
Storage2UriMapperJavaImpl is waiting for workspace lock which is owned by `"Worker-22: Updating workspace"` thread.

**Comment by cdietrich** [comment]
hmmmm. again all the mess about listing to jdt, storage to uri mapping and loading the index we dont have the time to work on

**Comment by cdietrich** [comment]
i wonder if a two phase lock would be better here
(1) i want to load the state
(2) i want to load the state and i have already a workspace lock

**Comment by szarnekow** [comment]
Fixed by [xtext-eclipse#1083]
